### ISSUE TITLE
Parse Spark errors from logs

### ISSUE THREAD

**Issue by davidmarin** [Yelp/mrjob#2056]
Right now you have to sift through Spark driver output.

**Comment by davidmarin** [comment]
It looks like this is a general problem with mrjob running spark-submit; we just echo the output. See `_run_spark_submit` in `mrjob/bin.py`.

**Comment by davidmarin** [comment]
Seems like the most helpful thing would be to find the Python exception. It looks something like this:

```
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564461935490_0001/container_1564461935490_
0001_01_000002/pyspark.zip/pyspark/worker.py", line 230, in main
    process()
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564461935490_0001/container_1564461935490_
0001_01_000002/pyspark.zip/pyspark/worker.py", line 225, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564461935490_0001/container_1564461935490_
0001_01_000002/pyspark.zip/pyspark/serializers.py", line 139, in dump_stream
    for obj in iterator:
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564461935490_0001/container_1564461935490_
0001_01_000001/pyspark.zip/pyspark/rdd.py", line 1573, in func
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564461935490_0001/container_1564461935490_
0001_01_000002/pyspark.zip/pyspark/util.py", line 55, in wrapper
    return f(*args, **kwargs)
  File "mr_sparkaboom.py", line 28, in kaboom
    raise Exception('KABOOM')
Exception: KABOOM
```

**Comment by davidmarin** [comment]
It also looks like it's not a great idea to parse `stdout` before `stderr` when diagnosing Spark steps on EMR. You end up with an error like:

```
Traceback (most recent call last):
  File "mr_sparkaboom.py", line 42, in <module>
    MRSparKaboom.run()
  File "/usr/lib/python3.4/dist-packages/mrjob/job.py", line 446, in run
    mr_job.execute()
  File "/usr/lib/python3.4/dist-packages/mrjob/job.py", line 470, in execute
    self.run_spark(self.options.step_num)
  File "/usr/lib/python3.4/dist-packages/mrjob/job.py", line 689, in run_spark
    spark_method(input_path, output_path)
  File "mr_sparkaboom.py", line 36, in spark
    kaboomed_lines.saveAsTextFile(output_path)
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564516520174_0001/container_1564516520174_0001_02_000001/pyspark.zip/pyspark/rdd.py", line 1585, in saveAsTextFile
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564516520174_0001/container_1564516520174_0001_02_000001/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1257, in __call__
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564516520174_0001/container_1564516520174_0001_02_000001/py4j-0.10.7-src.zip/py4j/protocol.py", line 328, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o66.saveAsTextFile.
: org.apache.hadoop.mapred.FileAlreadyExistsException: Output directory s3://mrjob-35cdec11663cb1cb/tmp/mr_sparkaboom.marin.20190730.195257.296911/output already exists
```

when the actual error is something like:

```
Caused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564516520174_0001/container_1564516520174_0001_01_000002/pyspark.zip/pyspa
rk/worker.py", line 230, in main
    process()
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564516520174_0001/container_1564516520174_0001_01_000002/pyspark.zip/pyspa
rk/worker.py", line 225, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564516520174_0001/container_1564516520174_0001_01_000002/pyspark.zip/pyspark/serializers.py", line 139, in dump_stream
    for obj in iterator:
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564516520174_0001/container_1564516520174_0001_01_000001/pyspark.zip/pyspark/rdd.py", line 1573, in func
  File "/mnt/yarn/usercache/hadoop/appcache/application_1564516520174_0001/container_1564516520174_0001_01_000002/pyspark.zip/pyspark/util.py", line 55, in wrapper
    return f(*args, **kwargs)
  File "mr_sparkaboom.py", line 28, in kaboom
    raise Exception('KABOOM')
Exception: KABOOM
```

This error is actually found in both files, but `stderr` only contains the exception from the executor (the proximate cause), while `stdout` starts with the error from the driver.

The full error is actually a chain of Java and Python exceptions, but that's not actually useful to the user; they can scroll up and read the log themselves if they want that.

Basically, Spark error log parsing is just a different beast.

**Comment by davidmarin** [comment]
It looks like basically anything that's has log level `ERROR` is potentially useful for determining why a job failed. To pick the best error, I'd look for:

* contains a Python stack trace
* shortest (but not a single line)
* earliest?

**Comment by davidmarin** [comment]
Spark mixes java and Python stack traces in the same message, so instead of generalizing `hadoop_error` to `java_error`, I think we should just introduce `spark_error` as a possible error type.

**Comment by davidmarin** [comment]
When running on a `local-cluster[*]` master, sometimes `WARN` messages contain better errors.

**Comment by davidmarin** [comment]
Parsing the driver log is not useful for cluster mode (e.g. on EMR), where we have to parse container logs. Still, using the Spark parser would be better than the one we have designed for Hadoop.

We might be able to get away with just parsing the log for the container that contains the driver, which is good because there doesn't seem to be a history log to point us towards the right files. Parsing from the last container tends to dig up irrelevant errors (e.g. output directory already full).

In any case, handling cluster mode might be better as a separate ticket.

**Comment by davidmarin** [comment]
If there's a Python Traceback, it might be more useful to extract just that when parsing error logs. Including the Java traceback, etc. might be rather verbose. For example, here's a traceback from running a broken job on a `local-cluster` master:

```
Probable cause of failure:

Lost task 1.0 in stage 0.0 (TID 1, 10.0.0.203, executor 2): org.apache.spark.SparkException: Task failed while writing rows
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:155)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:83)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:78)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:402)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:408)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Users/davidmarin/git/venv-2.7-mrjob/lib/python2.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 372, in main
    process()
  File "/Users/davidmarin/git/venv-2.7-mrjob/lib/python2.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 367, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/Users/davidmarin/git/venv-2.7-mrjob/lib/python2.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/serializers.py", line 141, in dump_stream
    for obj in iterator:
  File "/Users/davidmarin/git/venv-2.7-mrjob/lib/python2.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 1558, in func
  File "/Users/davidmarin/git/venv-2.7-mrjob/lib/python2.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/util.py", line 99, in wrapper
    return f(*args, **kwargs)
  File "/Users/davidmarin/git/mrjob/mrjob/examples/mr_sparkaboom.py", line 28, in kaboom
    raise Exception('KABOOM')
Exception: KABOOM

	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)
	at org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)
	at org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)
	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:128)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:127)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1394)
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:139)
	... 10 more
```

**Comment by davidmarin** [comment]
For starters, going to get this working in the Spark and local runners, which just call `_run_spark_submit()`.

Next, should get this working on EMR, where we're always in cluster mode and we need to parse container logs. It might be as simple as just parsing the log for the container that contains the driver.

Finally, we should combine the knowledge above to make this work on the Hadoop runner (have `_run_job_in_hadoop()` use `_run_spark_submit()` for Spark steps, parse container logs only if we're running in cluster mode).

**Comment by davidmarin** [comment]
Just trimming out the part of the Spark error *after* the Python traceback would be helpful, since the main goal is to keep people from having to scroll up.
